stac_api/clients/postgres/session.py

```diff
@@ -0,0 +1,53 @@
+"""database session management"""
+import logging
+import os
+from contextlib import contextmanager
+from typing import Iterator
+
+import attr
+import sqlalchemy as sa
+from sqlalchemy.orm import Session as SqlSession
+
+import psycopg2
+from fastapi_utils.session import FastAPISessionMaker as _FastAPISessionMaker
+from stac_api import errors
+
+logger = logging.getLogger(__name__)
+
+
+class FastAPISessionMaker(_FastAPISessionMaker):
+    """FastAPISessionMaker"""
+
+    def __init__(self, database_uri: str):
+        """init"""
+        super().__init__(database_uri)
+
+    @contextmanager
+    def context_session(self) -> Iterator[SqlSession]:
+        """override base method to include exception handling"""
+        try:
+            yield from self.get_db()
+        except sa.exc.IntegrityError as e:
+            if isinstance(e.orig, psycopg2.errors.UniqueViolation):
+                raise errors.ConflictError("resource already exists") from e
+            elif isinstance(e.orig, psycopg2.errors.ForeignKeyViolation):
+                raise errors.ForeignKeyError("collection does not exist") from e
+            logger.error(e, exc_info=True)
+            raise errors.DatabaseError("unhandled database error")
```

Comment:
Do we need a catch-all exception handler here?
```suggestion
            raise errors.DatabaseError("unhandled database error")
       except Exception as e:
            logger.error(e, exc_info=True)
            raise errors.DatabaseError("unhandled database error")
         
```

Reply:
yes good catch, we need to handle exceptions from `get_db` which are not of `IntegrityError`.

Reply:
catch-all exception handler doesn't work here because there are other exceptions that might get thrown within the context of a database session (like `NotFoundError`).  These exceptions end up getting lost in the catch-all exception handler which then impacts response codes.

Instead I've updated the `except` to use a lower-level sqlalchemy exception.  It's the lowest level that includes the wrapped psycopg2 error (`e.orig`) that we need to determine what error to throw.  It should cover [any error which occurs during the execution of a SQL statement](https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/exc.py#L377-L387), which I think covers our requirements here.